weave/trace_server/clickhouse_trace_server_batched.py

```diff
@@ -766,118 +774,104 @@ def add_new_row_needed_to_insert(row_data: Any) -> str:
-                            )
-                        WHERE rn = 1
-                        ORDER BY project_id, digest
-                    )
-                    ARRAY JOIN row_digests AS row_digest
-                    WHERE digest = {{digest:String}}
-                ) AS t
-                JOIN (
-                    SELECT project_id, digest, val_dump
-                    FROM (
-                            SELECT *,
-                                row_number() OVER (PARTITION BY project_id, digest) AS rn
-                            FROM table_rows
-                            WHERE project_id = {{project_id:String}}
-                        )
-                    WHERE rn = 1
-                    ORDER BY project_id, digest
-                ) tr ON t.project_id = tr.project_id AND t.row_digest = tr.digest
-                WHERE {predicate}
-                {sort_clause}
-            """
-        if parameters is None:
-            parameters = {}
-        if limit:
-            query += " LIMIT {limit: UInt64}"
-            parameters["limit"] = limit
-        if offset:
-            query += " OFFSET {offset: UInt64}"
-            parameters["offset"] = offset
+        if sort_fields is None or len(sort_fields) == 0:
+            sort_fields = [
+                OrderField(
+                    field=QueryBuilderField(field=ROW_ORDER_COLUMN_NAME),
+                    direction="ASC",
+                )
+            ]
 
-        query_result = self.ch_client.query(
-            query,
-            parameters={
-                "project_id": project_id,
-                "digest": digest,
-                **parameters,
-            },
-        )
+        if (
+            len(sort_fields) == 1
+            and sort_fields[0].field.field == ROW_ORDER_COLUMN_NAME
+            and (sql_safe_conditions is None or len(sql_safe_conditions) == 0)
+        ):
+            query = make_natural_sort_table_query(
+                project_id,
+                digest,
+                pb,
+                limit=limit,
+                offset=offset,
+                natural_direction=sort_fields[0].direction,
+            )
+        else:
+            sql_safe_sort_clause = f"ORDER BY {', '.join([sort_field.as_sql(pb, TABLE_ROWS_ALIAS) for sort_field in sort_fields])}"
+            query = make_standard_table_query(
+                project_id,
+                digest,
+                pb,
+                sql_safe_conditions=sql_safe_conditions,
+                sql_safe_sort_clause=sql_safe_sort_clause,
+                limit=limit,
+                offset=offset,
+            )
+
+        query_result = self.ch_client.query(query, parameters=pb.get_params())
 
         return [
             tsi.TableRowSchema(digest=r[0], val=json.loads(r[1]))
             for r in query_result.result_rows
         ]
 
+    def table_query_stats(self, req: tsi.TableQueryStatsReq) -> tsi.TableQueryStatsRes:
+        parameters: Dict[str, Any] = {
```

Comment:
-> `req.model_dump()`

Reply:
i do not want to blindly dump for params - i like the explicit nature of this as i want full clear control over the params sent to clickhous

Reply:
You can do `model_dump(include={'project_id', 'digest'})`

But you are right -- in this case where the dict is small, just making your own dict is also more performant